Some meta-attribute best practices

nrwahl2 opened the pull request:
Mostly new/renamed constants and deprecations, so there are a lot of commits but most are trivial.

nrwahl2 commented on lib/pengine/unpack.c:
```diff
@@ -542,20 +542,26 @@ expand_remote_rsc_meta(xmlNode *xml_obj, xmlNode *parent, pcmk_scheduler_t *data
+                continue;
+            }
+
+            if (strcmp(name, PCMK__META_REMOTE_NODE) == 0) {
```
The `pcmk__str_eq()` calls became two-line with the new string constants

nrwahl2 commented on include/crm/msg_xml.h:
```diff
@@ -219,13 +219,19 @@ extern "C" {
+/* @TODO These are op attrs that are never set anywhere. Evaluate whether they
+ * (and the code added with them in 5ddfc16f) can be dropped.
+ *
+ * If kept, they should be renamed to reflect that they're op attrs.
```
Now that I know these are for T382, only renaming them remains to be done.

That raises a question that I've been pondering: do we want to stick with our current naming approach? For reference, that is:
* `PCMK_XE_*`
* `PCMK_XA_*`
* `PCMK__META_*`

There are two subquestions here:
1. Do we want to stick with public names? Obviously I've gone with internal names for `PCMK__META_*` in this PR, because there was precedent prior to (public) `PCMK_META_ENABLED`. If so, what should be public and what should be internal? It might be reasonable for element names -- perhaps only CIB element names -- to be public. It's harder to envision public use cases for XML attributes, and even harder to envision public use cases for meta-attribute names.
2. Are these prefixes sufficiently descriptive and accurate?

To illustrate a couple of problems with #2:
* Some things (e.g., "enabled") can be used equally well as both XML attributes and meta-attribute names.
  * Used only as XML attribute for operations (not yet supported as a meta-attribute there, though it should be)
  * Used only as meta-attribute for alerts
* Operations are documented as accepting either element attributes or meta-attributes. This means the same constant should be both `PCMK_XA_*` and `PCMK__META_*`.
  * To a lesser extent this is true for resources as well, although I don't think it's documented. Resource element attributes (e.g., `provider` and `type`) get added to the meta-attribute hash.

You can also make an argument that more "semantic" naming is clearer, by allowing grouping of all resource attributes, all operation attributes, all alert attributes, etc. This is basically what the old naming scheme does. It's kind of nice to see be able to know immediately both:
* "this is the set of attributes that will be used with object type X", and
* "this particular attribute will definitely be used with object type X (so that if I'm doing something with object type X's options, I won't overlook this one)"

Semantic naming allows us to have `PCMK__ALERT_ATTR_PATH` and `PCMK__ALERT_ATTR_ENABLED` (or `PCMK__ALERT_META_ENABLED`) -- they're used in different positions (XML attribute vs. meta-attribute) but are both for alerts.

It also allows us to distinguish from a large list of meta-attributes, which ones are for which object types.

---

The planned grouping of option metadata into comprehensive arrays with object type indicated, addresses some of this but not all of it.

kgaillot replied:
My inclination is:
* Symbols should be public if the user may specify them somewhere (especially the CIB), internal if they're only used internal to Pacemaker (such as daemon IPC/CPG message XML)
* I usually favor something more descriptive than XE/XA (for XML elements/attributes), but we have so many XML-related APIs that anything longer gets unwieldy
* The same literal string may have more than one constant for different contexts. For example if something can be an XML element name in one context and an enumerated option value in another. For meta-attributes that can be specified as either XML attributes or nvpair names, it's borderline -- using just the meta constant would be fine since the XML attribute is still in a meta-attribute context, but having two constants would also be fine for consistency. I don't think we need to distingiush different meta-attribute contexts (such as resource vs alert).

I expect the main categories we would need are:
* XA = XML attribute name
* XE = XML element name
* META = meta-attribute
* OPT = cluster option
* VALUE = enumerated value of an XML attribute, name/value pair, meta-attribute, cluster option etc.

nrwahl2 replied:
Sounds good :)

> using just the meta constant would be fine since the XML attribute is still in a meta-attribute context, but having two constants would also be fine for consistency

"just the meta constant" is my general preference there. We're usually only going to check for the name once, rather than checking separately for an XML attribute or a meta-attribute. (There are some exceptions, like "enabled" for operations -- we check only for an XML attribute currently, but hopefully that changes in the future.)

kgaillot commented on agents/ocf/remote.in:
```diff
@@ -28,11 +28,11 @@ meta_data() {
   <shortdesc lang="en">Pacemaker Remote connection</shortdesc>
   <parameters>
-    <parameter name="server" unique-group="address">
+    <parameter name="addr" unique-group="address">
```
Bah, I never noticed this discrepancy. "server" is what users have been seeing (from the metadata, which is used by both pcs and the Pacemaker Remote document). We have 10 regression tests using "server" and none using "addr" as a result. Unfortunately we can't change the "addr" entries added by Pacemaker without changing the config hash.

Also unfortunately, they're not exactly equivalent. pe__bundle_needs_remote_name(), pcmk__substitute_remote_addr(), and add_replica_actions_to_graph() via XPATH_REMOTE, only check "addr". (Using an XPath to search for a meta-attribute is problematic in general due to the possibility of multiple blocks using rules. That wouldn't make sense for this particular meta-attribute, but strange configs happen.) At least we could fix these spots to check for both.

I tend to think we should keep "server" since it is the only one that has been documented all this time. We could maybe change "addr" at 3.0.0.

nrwahl2 replied:
> Unfortunately we can't change the "addr" entries added by Pacemaker without changing the config hash.

Yep that was the reason I chose "addr" over "server".

> Also unfortunately, they're not exactly equivalent. pe__bundle_needs_remote_name(), pcmk__substitute_remote_addr(), and add_replica_actions_to_graph() via XPATH_REMOTE, only check "addr".

I don't think this matters. From a skim, those functions are looking for "#uname" as a special value of "addr", and only Pacemaker would ever set that. If a user sets `addr="#uname"`, that's a grave mistake and a nonsensical configuration AFAIK.

Also, if there is a situation where the user sets "server=<some_address>" and (for some reason) Pacemaker sets "addr" to "#uname" and then substitutes a calculated address later, the result in the RA would be undefined. The last one in the param list ("addr" or "server") takes precedence in `handle_remote_ra_start()`.

> we could fix these spots to check for both

No need, checking for `addr="#uname"` is checking for a value that Pacemaker planted there as part of the remote container hack. We only need to check whichever variable Pacemaker uses for the hack.

> I tend to think we should keep "server" since it is the only one that has been documented all this time. We could maybe change "addr" at 3.0.0.

We also use "remote-addr" (not "remote-server") as a guest node resource meta-attribute. In that case, we pass its value as the "addr" (not "server") of the implicit remote connection resource (via `pe_create_remote_xml()`).

The remote RA metadata is the **only** place we use "server". Deprecating it in favor of "addr" seems reasonable to me.

Waiting for 3.0.0 and then going the other direction -- changing all uses to "server" instead of changing the metadata to "addr" -- is still an option too.

kgaillot replied:
> The remote RA metadata is the **only** place we use "server".

Sure, but that's the documentation, so that's what users have been using

> Waiting for 3.0.0 and then going the other direction -- changing all uses to "server" instead of changing the metadata to "addr" -- is still an option too.

I'm thinking that keeping the current situation is the least disruptive.

nrwahl2 replied:
[T743](https://projects.clusterlabs.org/T743)

nrwahl2 commented on lib/pengine/complex.c:
```diff
@@ -752,7 +752,7 @@ pe__unpack_resource(xmlNode *xml_obj, pcmk_resource_t **rsc,
 
     value = g_hash_table_lookup((*rsc)->meta, XML_RSC_ATTR_MANAGED);
-    if (value != NULL && !pcmk__str_eq("default", value, pcmk__str_casei)) {
+    if (value != NULL) {
```
We can definitely do that. I wouldn't have bothered in this case simply because there's no reason anyone should have known about it or used it. "Never documented" could reasonably mean "no contract."

nrwahl2 replied:
Created T742 if we must keep it :)

nrwahl2 commented on include/crm/msg_xml.h:
```diff
@@ -50,10 +50,9 @@ extern "C" {
+/* This has been deprecated as a CIB <clone> element attribute (alias for
+ * "promoted-node-max") since 2.0.0.
  */
-#define PCMK_XA_PROMOTED_MAX_LEGACY         "master-max"
```
I mention that here, as one of my concerns with `XA/META` naming vs. semantic, context-based naming: https://github.com/ClusterLabs/pacemaker/pull/3301#discussion_r1439281805

It's especially relevant for operation attributes. Operations are explicitly documented as accepting their attributes as either XML attributes (of the `op` element) or as meta-attribute `nvpair`s. So neither `XA` nor `META` is correct -- or, both are correct.

nrwahl2 commented on include/crm/common/options_internal.h:
```diff
@@ -133,6 +133,7 @@ bool pcmk__valid_sbd_timeout(const char *value);
 
 // Constants for meta-attribute names
 #define PCMK__META_ALLOW_UNHEALTHY_NODES    "allow-unhealthy-nodes"
+#define PCMK__META_CLONE_MAX                "clone-max"
```
As long as we're consistent that's fine by me. My thought was "default to internal" for things without a clear public use case... which per https://github.com/ClusterLabs/pacemaker/pull/3301#discussion_r1439281805 includes all meta-attributes, and arguably XML attributes and (much less convincingly) XML elements.

To your point, down the road someone might want to update meta-attributes using the public API.

kgaillot replied:
Yeah my thinking is that someone might want to auto-generate a CIB, so any literal strings that could be used in a CIB should be public (XML attribute and element names, meta-attribute names, cluster option names, and enumerated values). It's fine to have some public XML element names and some internal (such as those used only in IPC/CPG messages).

nrwahl2 commented on lib/pengine/common.c:
```diff
@@ -505,15 +505,17 @@ add_hash_param(GHashTable * hash, const char *name, const char *value)
-        return;
 
-    } else if (pcmk__str_eq(value, "#default", pcmk__str_casei)) {
+    if (name == NULL || value == NULL) {
```
Added to the description of T742.

nrwahl2 commented:
Updates:
* Dropped the "default" commits pending T742.
* Dropped the remote RA "server" commit pending T743.
* Made symbols public if they might be set by a user.
* Updated "clear_failure_op/interval" commit based on knowledge of T382.
* Added a commit clarifying the naming conventions.

Will rebase on current main after review.

kgaillot reviewed (commented):
Looks good. I think the meta/instance attribute constants should go in the new crm/common/options.h rather than msg_xml.h (which would end up with just XE/XA constants).

nrwahl2 commented:
I'm gonna try to get T711 merged first (and then rebase on main) since there will be more conflicts anyway. That one will introduce `options.h`.

nrwahl2 commented:
Rebased on current main and added two commits on top to move symbols to options.h and use them in more log messages, comments, etc. (not really necessary but meh)

nrwahl2 commented:
This will need one more rebase if #3303 (which should finish building first) gets merged first.

nrwahl2 commented:
Rebased on current main and resolved conflicts. No other changes.
